### ISSUE TITLE
3D building:part different levels are not rendered correctly

### ISSUE THREAD

**Issue by famo** [OSMBuildings/OSMBuildings#163]
According to https://wiki.openstreetmap.org/wiki/Key:building:part

> Where a building has been split into parts, the whole building way should still have the tags building:levels=* and height=*. **These values should be the maximum of all the building parts.** A building that has a three-story part and a six-story part is still a six-story building.

However buildings tagged this way are not rendered correctly in OSMB, see example:
https://osmbuildings.org/?lat=51.35417&lon=7.48997&zoom=19.8&tilt=18&rotation=-41
Correct rendering:
https://demo.f4map.com/#lat=51.3542371&lon=7.4899911&zoom=20&camera.theta=24.71&camera.phi=47.842

**Comment by kekscom** [comment]
This object is a combination of a building part, largely intersected by a building.
Either the intersection should be resolved or both parts should be tagged appropriately: building:part

**Comment by kekscom** [comment]
Addition: this was hard to spot. It is a building, enclosing 2 building:parts.
It is correct, not to show the building. But one of the parts is missing.
Hence that is the issue which has to be fixed.

**Comment by famo** [comment]
@kekscom Your first comment / observation was correct. I fixed it in OSM in the meantime, and for the building displays accordingly now.
IMHO this can be closed...

**Comment by mimi89999** [comment]
Hello,
I have the same issue with https://www.openstreetmap.org/way/764472237#map=19/52.22202/21.00663 rendered incorrectly in https://osmbuildings.org/?lat=52.22216&lon=21.00702&zoom=19.4&tilt=40&rotation=27

@kekscom Do I understand correctly from your comments, that all parts of a building should be marked as parts and tagged? That's not how I understand the doc:

> A way with the tag building:part=* describes a part of a building where some building attributes or the building function are different from the rest of the building.

> Use building:part=yes for parts of the building which only have different attributes (building:levels=* and height=*).

I understand that only parts of the building where attributes have a different value should be tagged.

**Comment by kekscom** [comment]
Hi,
You're simply nesting a part into a building without further indication how to handle ist. It seem rather positive that no element has been discarded so far.

> I understand that only parts of the building where attributes have a different value should be tagged.

I've never seen that. Also implementation would be horrible.

I recommend to split into 3 parts (vertically or horizontally). If you'd like to place common tags add them all to a relation type=building.

**Comment by mimi89999** [comment]
>> I understand that only parts of the building where attributes have a different value should be tagged.

> I've never seen that. Also implementation would be horrible.

Isn't it what the doc says?

> You're simply nesting a part into a building without further indication how to handle it.

I think that this situation can be interpreted in only one way.

Covering the entire building with parts would add a lot of redundant data. If a building has `n` parts connected to a central part, then it would have `n + 1` parts. That's a lot. Imagine a building in the form of a triangle with a nested triangle or four squares attached to the corners of a fifth square.

**Comment by kekscom** [comment]
> Covering the entire building with parts would add a lot of redundant data. If a building has n parts connected to a central part, then it would have n + 1 parts. That's a lot. Imagine a building in the form of a triangle with a nested triangle or four squares attached to the corners of a fifth square.

This is how everyone does it. Can't see redundancy here.
For more detailed tagging discussion OSM mailing lists would be a good be a good place.

**Comment by mimi89999** [comment]
I changed that building: https://www.openstreetmap.org/relation/10613373
I hope it's OK now.

**Comment by kekscom** [comment]
You could add `levels` or `height` to every part.
The additional relation is fine for grouping the building but not urgently necessary unless there are common attributes like address, building name etc.

**Comment by mimi89999** [comment]
Now it's even more broken than before and I don't understand why.

**Comment by mimi89999** [comment]
@kekscom Now the building renders very incorrectly. Do I need to add levels and height to every part?

**Comment by kekscom** [comment]
Yes height/levels should be attached to every part.
I've just done that. let's see how it works.

**Comment by mimi89999** [comment]
@kekscom  It looks much better, but now the part that is removed seems only to be half a floor high, and floors above it appear higher that the rest.

**Comment by natrius** [comment]
May i join here? 

I have a building https://osmbuildings.org/?lat=47.82872&lon=13.03745&zoom=19.0&tilt=36&rotation=27 

![grafik](https://user-images.githubusercontent.com/7243447/75337331-138f1a80-588d-11ea-8874-584779dbf592.png)

And out of this one https://wiki.openstreetmap.org/wiki/Simple_3D_buildings there is this part:

> When a building has any building:part=* areas, the building outline is not considered for 3D rendering. 

From my pic it seems like the outline is rendered as well? Or am i missing something?

**Comment by kekscom** [comment]
Yes all 3 mapped objects are rendered. Because these are in no logical relation to each other.
Creating a relation might help. Where all parts are market as part and outline is marked as such.

Our intersection test might solve this but it only catches containment or 1:1 identity.

**Comment by natrius** [comment]
So the wiki is (at least for osmbuildings) not correct right now and you don't check if the building contains a `building:part=yes` and if==yes ignore the building=yes. Because of the wiki it lead to tagging the complete building with 

```
building=yes
building:part=yes
building:levels=3
```

and the inner raised part with 

```
building:part=yes
building:levels=4
```

to get 

![grafik](https://user-images.githubusercontent.com/7243447/75343045-e5fb9e80-5897-11ea-98a8-fb64d9a58931.png)

I will give relations a try, but from what i get now it could also help to just give the whole outline a `building:levels=0` or 1 or something like that. 

Thanks for your answer!

**Comment by natrius** [comment]
Actually, some mpas/apps seem to take this into account (the wiki part about ignoring the building:levels from the whole building if building:part=* is present). OpenStreetMap for example suggests to just put in the highest buildingpart for the complete building.

**Comment by natrius** [comment]
So, this whole building gets now rendered wrong https://www.openstreetmap.org/way/154860143 because of that. I don't want to put up a relation or some (wrong) workaround where i put a `building:levels=0` on the surrounding building. 

And i guess it should be detected by your intersection test, if understand it correctly, because it is a 1:1 identity. 

https://wiki.openstreetmap.org/wiki/Simple_3D_buildings

> When a building has any building:part=* areas, the building outline is not considered for 3D rendering.

Unfortunately i cannot help otherwiseright now, but i think this is a huge problem right now, because some people seem to have started tagging `building:levels=0` as a workaround and thats just the 'famous' "Tagging for the renderer".

**Comment by natrius** [comment]
If it helps, you could set up a bug-bounty, i'm happy to pay a little bit to help developing with this.

**Comment by natrius** [comment]
Oh, my bad, i meant just "it should be set up" and well, i did. https://www.bountysource.com/issues/69127579-3d-building-part-different-levels-are-not-rendered-correctly

Anybody who is solving this can claim the bounty.

**Comment by k-vde** [comment]
Hi all,
I'm having the same issue with https://osmbuildings.org/?lat=51.10571&lon=4.91596&zoom=19.8&tilt=29&rotation=17
I have building parts covering almost the entire building but the building itself still renders with the height I set for it. I currently added the wrong height of 3m while i try to sort out the other problems. Can someone tell me what I'm doing wrong?

**Comment by natrius** [comment]
You did nothing wrong. Its an osmbuildings issue. 

You could create a building-relation to resolve it a better way than setting a wrong height. But in my opinion that should not be neccessary.

**Comment by kekscom** [comment]
Agreed it looks like other issues.
However, what is the expected outcome?
I see the building (3m) and parts intersecting it. How should it be rendered?

**Comment by ghost** [comment]
it makes no sense not to show the building, when in osm there is a building

**Comment by natrius** [comment]
@kekscom 
Whopps, completely missed your answer, sorry. 

Example: https://osmbuildings.org/?lat=47.82936&lon=13.05276&zoom=19.6&tilt=32&rotation=-8
contains of a building without any values (https://www.openstreetmap.org/way/777739034) and 3 smaller ones(https://www.openstreetmap.org/way/800694347, https://www.openstreetmap.org/way/800694346 and https://www.openstreetmap.org/way/800694345) with the appropriate values, filling out the complete building. 

The main-problem is, you render a default-height for the building when a building:part is present. Remove that and a lot of problems are gone. That would also be with https://wiki.openstreetmap.org/wiki/Simple_3D_buildings#Building_outlines and specifically 

> When a building has any building:part=* areas, the building outline is not considered for 3D rendering. 

If(building contains (building:part && building:levels){do not render} or revers, of course.

(And yeah, i could solve this with an relation like i did a little bit in the east, but this is a great example :D )
